### PR TITLE
Optimize event serialization with pre-allocated buffer

### LINKED ISSUE #1082: TODO replace to_allocvec with an incremental buffer in the LLMP manager
Just noticed that we do useless allocations

### PR DESCRIPTION
**Pull request by mzfr**
Fixes #1082

I tested the changes with libpng fuzzer, and noticed quite a significant performance improvement, though I haven't done any formal benchmarking.

Current Implementation Issue:
The fuzzer panics after 4-5 minutes of execution with a "SerializeBuffer full" error. While increasing the initial buffer size could help, it might not be the most elegant solution(?)

Questions/Feedback Needed:
1. Should I implement a fallback mechanism to use `to_allocvec` when the pre-allocated buffer is full?
2. Any other way to handle the Buffer full issue?

### CODE REVIEW

**Comment by tokatoka** [comment]
> The fuzzer panics after 4-5 minutes of execution with a "SerializeBuffer full" error. While increasing the initial buffer size could help, it might not be the most elegant solution(?)

Yeah it's difficult... 
@domenukk any idea?

imo i don't think using pre-allocated buffer makes it significantly efficient, but on the other hand it makes the code unnecessarily complex.

**Comment by domenukk** [comment]
In the error case, can't we just use `to_allocvec`, and replace the current `event_buffer` with the newly allocated vec?

**Comment by tokatoka** [comment]
sounds good

File: libafl/src/events/llmp/mgr.rs [hunk 1]
```diff
@@ -199,6 +201,7 @@ impl<EMH> LlmpEventManagerBuilder<EMH> {
             time_ref,
             phantom: PhantomData,
             custom_buf_handlers: vec![],
+            event_buffer: Vec::with_capacity(1024 * 4),
```
Comment (domenukk) [inline libafl/src/events/llmp/mgr.rs:204]: Can we make the initial size a const?

File: libafl/src/events/llmp/mgr.rs [hunk 2]
```diff
@@ -512,44 +518,57 @@ where
...
         let flags = LLMP_FLAG_INITIALIZED;
 
-        match self.compressor.maybe_compress(&serialized) {
-            Some(comp_buf) => {
-                self.llmp.send_buf_with_flags(
-                    LLMP_TAG_EVENT_TO_BOTH,
-                    flags | LLMP_FLAG_COMPRESSED,
-                    &comp_buf,
-                )?;
+        self.event_buffer.clear();
```
Comment (domenukk) [inline libafl/src/events/llmp/mgr.rs:529]: Why clear?

File: libafl/src/events/llmp/mgr.rs [hunk 3]
```diff
@@ -512,44 +518,57 @@ where
...
+            Err(e) => return Err(Error::from(e)),
+        };
+
+        #[cfg(feature = "llmp_compression")]
+        {
+            match self
+                .compressor
+                .maybe_compress(&self.event_buffer[..written_len])
+            {
+                Some(comp_buf) => {
```
Comment (domenukk) [inline libafl/src/events/llmp/mgr.rs:549]: _technically_ this still allocates, so we have future room for improvements

Comment (mzfr) [reply libafl/src/events/llmp/mgr.rs:549]: Sorry, I don't understand what you mean exactly. Are you referring to the allocation happening in maybe_compress function?

Comment (domenukk) [reply libafl/src/events/llmp/mgr.rs:551]: Yes in theory this also could use a preallocated buffer

**Comment by domenukk** [comment]
Thank you! :)

**Pull request merged**